spec/requests/api/v1/allocation_events_controller_cancel_spec.rb

```diff
@@ -0,0 +1,148 @@
+      end
+
+      it 'updates the allocation moves cancellation reason' do
+        expect(allocation.reload.moves.pluck(:cancellation_reason).uniq).to contain_exactly('other')
+      end
+
+      it 'updates the allocation moves cancellation reason comment' do
+        expect(allocation.reload.moves.pluck(:cancellation_reason_comment).uniq).to contain_exactly('FUBAR')
+      end
+
+      describe 'webhook and email notifications' do
+        it 'calls the move notifier when updating the status' do
+          expect(Notifier).to have_received(:prepare_notifications).with(topic: move, action_name: 'update_status')
+        end
+      end
+    end
+
+    context 'without specifying cancellation reason comment' do
+      let(:cancel_params) do
+        {
+          data: {
+            type: 'cancel',
+            attributes: {
+              timestamp: '2020-04-23T18:25:43.511Z',
+              cancellation_reason: cancellation_reason,
+            },
+          },
+        }
+      end
+
+      it 'updates the allocation cancellation reason comment' do
+        expect(allocation.reload.cancellation_reason_comment).to eq('Allocation was cancelled')
+      end
+    end
+
+    context 'with a bad request' do
+      let(:cancel_params) { nil }
+
+      it_behaves_like 'an endpoint that responds with error 400'
+    end
+
+    context 'when not authorized' do
+      let(:access_token) { 'foo-bar' }
+      let(:detail_401) { 'Token expired or invalid' }
+
+      it_behaves_like 'an endpoint that responds with error 401'
+    end
+
+    context 'with a missing id' do
+      let(:allocation_id) { 'foo-bar' }
+      let(:detail_404) { "Couldn't find Allocation with 'id'=foo-bar" }
+
+      it_behaves_like 'an endpoint that responds with error 404'
+    end
+
+    context 'with an invalid CONTENT_TYPE header' do
+      let(:content_type) { 'application/xml' }
+
+      it_behaves_like 'an endpoint that responds with error 415'
+    end
+
+    context 'with validation errors' do
+      context 'with a bad timestamp' do
+        let(:timestamp) { 'Foo-Bar' }
+
+        it_behaves_like 'an endpoint that responds with error 422' do
+          let(:errors_422) { [{ 'title' => 'Invalid timestamp', 'detail' => 'Validation failed: Timestamp must be formatted as a valid ISO-8601 date-time' }] }
+        end
+      end
+
+      context 'with a bad cancellation reason' do
+        let(:cancellation_reason) { 'raining' }
+
+        it_behaves_like 'an endpoint that responds with error 422' do
+          let(:errors_422) { [{ 'title' => 'Unprocessable entity', 'detail' => 'Cancellation reason is not included in the list' }] }
+        end
+      end
+
+      context 'with a missing cancellation_reason' do
+        let(:cancellation_reason) { nil }
+
+        it_behaves_like 'an endpoint that responds with error 422' do
+          let(:errors_422) { [{ 'title' => 'Unprocessable entity', 'detail' => 'Cancellation reason is not included in the list' }] }
```

Comment:
In the move events stuff the params validator is checking that a cancellation reason is provided for a cancel event. I didn't add that, instead relying on the allocation model to validate that a cancellation reason is provided when the state is cancelled. That avoids adding validation in two places; perhaps the move params validator should do the same?

Reply:
I am a bit wary about pushing the validation of events too far down the stack to the point of running the event. We had a bug last week where an event describing an invalid object was successfully created and saved, and then run. On running it crashed as the event was invalid, but it meant that no new events could be added for the object as the invalid one broke the chain.

In this implementation the event creator and the event runner are wrapped in a transaction in the controller, so any errors will not be saved to the db, so all should be well. Nevertheless, it seems a bit unsafe to rely on the runner as the primary validator.

Reply:
Hmm.. it's an interesting point. As you say, it's safe in this implementation. My counter argument would be that it's not a great pattern to be replicating validation rules in two places (i.e. the model and the params validator). On balance I prefer this approach unless there's a strong preference for something else?

Reply:
I appreciate the point (and agree) that duplication of code is not a good thing. However, I don't think that we can safely skip validation at this level.

How about instead moving the relevant validations which are needed on the event and on the move/allocation into their own concern, e.g.

```
module MoveEvents
  module CancellationReasonValidator
    extend ActiveSupport::Concern
    included do
      validates :cancellation_reason, inclusion: { in: Move::CANCELLATION_REASONS }, if: -> { type == 'cancel' }
    end
  end
end
```

then in the move events `param_validator.rb` we can include it:
```
module MoveEvents
  class ParamsValidator
    include ActiveModel::Validations
    include CancellationReasonValidator
    ...
```

and likewise include it in the move model / wherever else it should be validated. That way there is only one validation rule and it can be used in multiple places. Will that do the job?

Reply:
Moving the validation to a shared concern and then including it in the params validator and model still means it's in two places so needs two sets of specs. I'm not convinced it's a great pattern and prefer that the bulk validation work lives in the model - particularly the more complex stuff that depends on business rules or other state of the model. I think it's appropriate to check the existence of specific parameters in a request, and that date formats look sensible but we shouldn't take that approach too far and try to do everything in the params validator.

It's safe to apply this approach within a database transaction as if something goes boom the data is left in a consistent state. I'll add a spec to ensure that no events get created.